LanguageFeatures/Patterns/patterns_lib.dart

```diff
@@ -71,80 +70,84 @@ class Unit {
   }
 
   @override
-  String toString() => value.toStringAsFixed(2);
+  String toString() => value.toStringAsFixed(2).replaceFirst(".00", "");
 }
 
 class Shape {
-  String _log = "";
-  Shape();
+  final void Function(String s)? logFunction;
+  const Shape([this.logFunction = null]);
 
-  void logger(String toAppend) {
-    _log = _log.isEmpty ? toAppend : "$_log$toAppend";
+  void _log(String toLog) {
```

Comment:
Perhaps this could be:
```suggestion
  void _log(String toLog) => logFunction?.call(toLog);
```